music21/musicxml/xmlToM21.py

```diff
@@ -5691,7 +5691,19 @@ def xmlStaffLayoutFromStaffDetails(
                 stl.staffType = stream.enums.StaffType(xmlText)
             except ValueError:
                 environLocal.warn(f'Got an incorrect staff-type in details: {mxStaffType}')
+
         # TODO: staff-tuning*
+        mxStaffTuning = mxDetails.findall('staff-tuning')
+        if mxStaffTuning is not None:
+            tuning_pitches = []
+            for i in range(len(mxStaffTuning)):
+                staff_tuning = mxStaffTuning[i]
+                tuning_step = staff_tuning.find('tuning-step').text
+                tuning_octave = int(staff_tuning.find('tuning-octave').text)
+                tuning_pitches.append(pitch.Pitch(tuning_step + str(tuning_octave)))
+            fretboard = tablature.FretBoard.getFretBoardFromTuning(tuning_pitches)
+            setattr(stl, 'fretboard', fretboard)
```

Comment:
> But therefore I can not drop the identified fretboard in the beginning of the part as you suggested, so I tried to simply add it as a new attribute in the StaffLayer.

Can we just insert into the current measure? `staffLayoutToXmlStaffDetails` is a method on the MeasureExporter, so we have access to `self.stream`.

Reply:
`self.stream.append(fretboard)` does not work because `fretboard` is not a music21 object (the err message recommends to create a `music21.ElementWrapper`, should we ?)

Reply:
Goodness, alright. My mistake. Looking now, what do you suppose about setting `.stringPitches` on `self.activeInstrument`, if it is a `StringInstrument`? We don't even need to make or check fretboards, I guess.

Reply:
This is probably worth a ping for @mscuthbert before you end up wasting too much motion! Myke, any thoughts about these three options:

1a -- use FretBoards, and put them on StaffLayout objects
1b -- use FretBoards, but make them Music21Objects so they can go in streams
2 -- use `Instrument.stringPitches`

Reply:
One benefit of the instrument approach is that if we have ambiguous data for an instrument but we have staff-tuning tags, we could possibly reclassify the object to be a `StringInstrument`, or possibly even more specific.

Reply:
Thanks for the suggestions. For my part, using `self.activeInstrument` seems the most relevant. Of course this would require to identify the instrument of the part as guitar (I'm not sure how it's done but in case it might help I'm dropping here the musicXML score-part block of an [example tab file provided by MusicXML](https://www.w3.org/2021/06/musicxml40/musicxml-reference/examples/tutorial-tablature/) ):

```
      <score-part id="P2">
         <part-name print-object="no">Guitar [TAB]</part-name>
         <part-abbreviation print-object="no">Gtr.</part-abbreviation>
      </score-part>
```

The reason why I'm putting this here is that I'm a little bit afraid that the absence of `score-instrument` or `midi-instrument` tag could limit to identify the instrument as Guitar. Unless we can identify it from `part-name` ?

I think the reason why instrument informations are missing in part P2 is because they are already included in part P1 - Part P2 here is only intended to provide additional informations (fret, strings, etc.) on part P1 as it is usually the case in tablature notation (2 stacked scores telling the same, the first with notes the second one with fret/strings). But in term of instrument and track, P1 and P2 are one single part. I don't know if there is any way to link the 2 parts such that P2 will refer to the instrument of P1. I guess this goes beyond the current issue.

Reply:
> I don't know if there is any way to link the 2 parts such that P2 will refer to the instrument of P1.

In musicxml both staves should be within the same part and the second staff should have a staff-details `<staff-type>alternate</staff-type>` tag.

This is not something that music21 supports right now, but it's something we want to do, so test files with an implementation would be helpful.

https://www.w3.org/2021/06/musicxml40/musicxml-reference/examples/staff-type-element/

Reply:
If that's the ideal MusicXML representation, then I would think the tablature tutorial @louisbigo linked to should be updated to conform to it.

In https://github.com/w3c/musicxml/issues/205 they confirmed they wanted to move in this direction and update the documentation, so they would probably welcome a new issue for it.

That issue is also useful for its summary of the state of play.

Reply:
> In musicxml both staves should be within the same part and the second staff should have a staff-details <staff-type>alternate</staff-type> tag.

In that case, then these MusicXML documents are clearly representing optional notation for a single part, and then I think we can adopt @louisbigo's initial proposal of putting FretBoards on the `StaffLayout` object, possibly mixing in `StyleMixin` to FretBoard so that one can set `hideObjectOnPrint` False to suppress the alternate notation.

My thought of updating the instrument's stringPitches seems doable, but perhaps should be done in a separate PR.

> we shouldn't couple the fretboards too tightly to m21 StaffDetails objects just because MusicXML couples them (what about other formats, etc.)

Just to reiterate, this discussion is helping me no longer be as worried about this.

Reply:
I don't know that Fretboards belong on the StaffLayout object -- they're not about the layout of the staff, and if a fretboard gets moved to a different location it shouldn't change how the staff around it is laid out (unless it needs extra space for it; but the same could be said for notes on ledger lines, etc.)  -- they seem to me to be a kind of Harmony object subclass, like ChordSymbols.

Reply:
> they seem to me to be a kind of Harmony object subclass, like ChordSymbols.

Well, for _that_ we already have `tablature.ChordWithFretBoard`, but I didn't take that to be relevant here because we don't have chord symbols at all.

> I don't know that Fretboards belong on the StaffLayout object -- they're not about the layout of the staff, and if a fretboard gets moved to a different location it shouldn't change how the staff around it is laid out

Now I fear I've confused things. I think the issue Louis set out to solve really _is_ about the layout of the staff, i.e. alternate notation on a tab staff, not fretboard symbols that can be "moved" (what would it mean for staff lines of a tablature staff to "move"). But please do correct me if I'm off base here.

Reply:
I think that I'm a bit confused also.  We can have ChordWithFretBoard, but we can also have the same thing without a Chord symbol.  That's what I'm talking about that I call a FretBoard.  I think I'll need to see an image of what everyone is talking about to catch up.  But it doesn't seem like any of this belongs on StaffLayout unless it affects the size or amount of space around a staff or the change in the visual appearance.  If it's about the tuning of different lines, that's a different object.  I don't want to put things into one object that do not belong with it.

Reply:
The `FretBoard` definition in the music21 doc says : _A FretBoard represents a displayed fretboard (i.e. used in chord symbols)._. So I think that the `FretBoard` class has initially be thought to handle the displaying of string/frets positions accompanying chord symbols (see image below).


![freatboard](https://user-images.githubusercontent.com/57365368/143418211-cd029e0f-47ef-4fd3-805f-d887485ff9fe.png)

So the question is whether the class `FretBoard` could also be used to describe the string instrument linked to a whole tablature *staff*. Looking at the code of `FretBoard`, it seems to me that it is too related to *positions* (see for examples methods `getFretNoteByString` which does not make sense in the context of a whole tablature staff). So I'm starting to think that we might not want to use `Fretboard` here (I'm also wondering if to avoid futur similar confusion, we could consider renaming the class `FretBoard` in something like `FretBoardPosition`, but it would require changing the name of many related classes / methods with similarly ambiguous names in `tablature.py` like `FretNote`, etc.).

On the other hand, I agree that the tuning of the string instrument is not supposed to affect the layout of the staff so using `StaffLayout` might not seem justified here. Unfortunately it does not seem that  `<staff-tuning>` , as a `<staff-details>` child, could be easily be processed outside `xmlStaffLayoutFromStaffDetails`. Any idea of where it could be processed ?

Reply:
> So I'm starting to think that we might not want to use Fretboard here

I agree, we need some other data structure in music21 for tab staves. The image of what we're talking about can be found from the tutorial Louis linked to--we need to capture the pitches of the lines of a tab-staff, which may or may not (although it should) be in the same musicxml `<part>`:

![image](https://user-images.githubusercontent.com/38668450/143454733-1870027b-bfd6-4e5e-a756-30c8e83c6f44.png)


I'm thinking the `StaffType` enum needs a new member TAB, and StaffTypes do currently exist on StaffLayout objects. Myke, what do you think of that?

Where to encode the actual pitches, I'm not sure. We could add attributes to `TabClef` or we could keep `StringInstrument.stringPitches` up to date with the pitches of the tab staff. 🤷🏻‍♂️ 

> Any idea of where it could be processed ?

`handleStaffDetails()` could call a new method to read the `<staff-tuning>` tag.

Reply:
Maybe this means breaking out into two PartStaff objects when both are in one `<part>` just as we do for piano music.

Reply:
> we need to capture the pitches of the lines of a tab-staff, which may or may not (although it should) be in the same musicxml `<part>`

I think the tuning pitches will allways appear on the tab-staff (unless the file is uncorrect) - however the instrument information (guitar in that case) that might only appear in an other musicxml `<part>`.

> Where to encode the actual pitches, I'm not sure. We could add attributes to `TabClef` or we could keep `StringInstrument.stringPitches` up to date with the pitches of the tab staff.

Both make sense to me. But I guess using `TabClef` is more safe as `StringInstrument.stringPitches` might not work correctly in case instrument / parts are not indicated properly (problem above). So I'll go for `TabClef` is everyone is ok.

> `handleStaffDetails()` could call a new method to read the `<staff-tuning>` tag.

Good for me !